### PR TITLE
Enable standard library assertions in CI

### PR DESCRIPTION
**Pull request by vittorioromeo**
Addresses #3098. I've chosen flags that enable unexpensive assertions (e.g. assert when dereferencing a smart pointer or `std::optional`).

### CODE REVIEW

**Comment by kimci86** [comment]
Can't we determine which definition is needed with our CMake scripts depending on the platform and compiler? Then there could be a single option instead of three.

**Comment by vittorioromeo** [comment]
> Can't we determine which definition is needed with our CMake scripts depending on the platform and compiler? Then there could be a single option instead of three.

We'd have to detect the standard library being used, not the platform or compiler. You could e.g. compile with GCC on a MacOS using either libstdc++ or libc++.

I'm not sure how to reliably detect the standard library with CMake

Alternatively, we could just define all of them at once with a generic `SFML_ENABLE_STDLIB_ASSERTIONS` option. It doesn't hurt to have unused defines. 

What do you think? @eXpl0it3r

**Comment by vittorioromeo** [comment]
BTW, here's an example of an assertion that fires after intentionally adding a bad `std::optional` dereference in `sf::Image` as a test:

```
C:/msys64/ucrt64/include/c++/14.1.0/optional:475: _Tp &std::_Optional_base_impl<sf::Image, std::_Optional_base<sf::Image>>::_M_get() [_Tp = sf::Image, _Dp = std::_Optional_base<sf::Image>]: Assertion 'this->_M_is_engaged()' failed.
```

**Comment by eXpl0it3r** [comment]
CMake magic is @ChrisThrasher's specialty 😄  
Not sure what CMake offers for detection.

But I'm certain he'd also prefer one option (if any at all).

**Comment by binary1248** [comment]
> CMake magic is @ChrisThrasher's specialty 😄

😮

Combine [`iso646`](https://en.cppreference.com/w/cpp/header/ciso646) with CMake's `try_compile` to see what headers get consumed.

**Comment by ChrisThrasher** [comment]
> > Can't we determine which definition is needed with our CMake scripts depending on the platform and compiler? Then there could be a single option instead of three.
> 
> We'd have to detect the standard library being used, not the platform or compiler. You could e.g. compile with GCC on a MacOS using either libstdc++ or libc++.
> 
> I'm not sure how to reliably detect the standard library with CMake
> 
> Alternatively, we could just define all of them at once with a generic `SFML_ENABLE_STDLIB_ASSERTIONS` option. It doesn't hurt to have unused defines.
> 
> Basically, this:
> 
> ```cmake
> option(SFML_ENABLE_STDLIB_ASSERTIONS "Enable standard library assertions" OFF)
> if(SFML_ENABLE_STDLIB_ASSERTIONS)
>     add_definitions(-D_GLIBCXX_ASSERTIONS=1) # see https://gcc.gnu.org/wiki/LibstdcxxDebugMode
>     add_definitions(-D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_EXTENSIVE) # see https://libcxx.llvm.org/Hardening.html
>     add_definitions(-D_DEBUG=1) # see https://learn.microsoft.com/en-us/cpp/c-runtime-library/debug?view=msvc-170
> endif()
> ```
> 
> What do you think? @eXpl0it3r

I'd much prefer one CMake option that enables all 3 at once and then amend our `dev` preset to enable this by default so that the CI scripts remain unchanged.

**Comment by vittorioromeo** [comment]
> I'd much prefer one CMake option that enables all 3 at once and then amend our `dev` preset to enable this single option so that the CI scripts remain unchanged.

@ChrisThrasher 

This is a much much better solution! I've made the change you requested. For now, I removed `_DEBUG` for MSVC as it was causing linker errors, and I need to explore the consequences of enabling the full-fledged MSVC debug mode prior to doing so.

I would still recommend accepting this change for libstdc++ and libc++ to benefit from the assertions ASAP, and add MSVC support (if reasonable) in a later PR.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -63,7 +63,7 @@ jobs:
         - platform: { name: Windows MinGW, os: windows-2022 }
           config: { name: Static Standard Libraries, flags: -GNinja -DSFML_USE_MESA3D=TRUE -DCMAKE_CXX_COMPILER=g++ -DSFML_USE_STATIC_STD_LIBS=TRUE }
         - platform: { name: Windows MinGW, os: windows-2022 }
-          config: { name: Static with PCH (GCC), flags: -GNinja -DSFML_USE_MESA3D=TRUE -DCMAKE_CXX_COMPILER=g++ -DBUILD_SHARED_LIBS=FALSE -DSFML_ENABLE_PCH=1 }
+          config: { name: Static with PCH (GCC), flags: -GNinja -DSFML_USE_MESA3D=TRUE -DCMAKE_CXX_COMPILER=g++ -DBUILD_SHARED_LIBS=FALSE -DSFML_ENABLE_PCH=1 -SFML_ENABLE_STDLIB_ASSERTIONS=FALSE } # disabling stdlib assertions due to false positive
```
Comment (ChrisThrasher) [inline .github/workflows/ci.yml:66]: What's the false positive?

File: CMakeLists.txt [hunk 2]
```diff
@@ -204,6 +204,12 @@ if(SFML_ENABLE_SANITIZERS)
     string(APPEND CMAKE_CXX_FLAGS " -fno-omit-frame-pointer -fno-sanitize-recover=all -fsanitize=undefined")
 endif()
 
+option(SFML_ENABLE_STDLIB_ASSERTIONS "Enable standard library assertions" OFF)
+if(SFML_ENABLE_STDLIB_ASSERTIONS)
+    add_definitions(-D_GLIBCXX_ASSERTIONS=1)                                   # see https://gcc.gnu.org/wiki/LibstdcxxDebugMode
+    add_definitions(-D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_EXTENSIVE) # see https://libcxx.llvm.org/Hardening.html
```
Comment (ChrisThrasher) [inline CMakeLists.txt:210]:
```suggestion
    add_compile_definitions(_GLIBCXX_ASSERTIONS=1)                                   # see https://gcc.gnu.org/wiki/LibstdcxxDebugMode
    add_compile_definitions(_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_EXTENSIVE) # see https://libcxx.llvm.org/Hardening.html
```
CMake recommends using `add_compile_definitions` in lieu of `add_definitions`.

Source: https://cmake.org/cmake/help/latest/command/add_definitions.html

File: CMakePresets.json [hunk 3]
```diff
@@ -10,7 +10,8 @@
         "CMAKE_EXPORT_COMPILE_COMMANDS": "ON",
         "SFML_BUILD_EXAMPLES": "ON",
         "SFML_BUILD_TEST_SUITE": "ON",
-        "SFML_WARNINGS_AS_ERRORS": "ON"
+        "SFML_WARNINGS_AS_ERRORS": "ON",
+        "SFML_ENABLE_STDLIB_ASSERTIONS": "ON"
```
Comment (ChrisThrasher) [inline CMakePresets.json:14]:
```suggestion
        "SFML_ENABLE_STDLIB_ASSERTIONS": "ON",
        "SFML_WARNINGS_AS_ERRORS": "ON"
```
I'd like to keep these in alphabetical order if you don't mind

**Review by ChrisThrasher — CHANGES_REQUESTED** [review]
LGTM pending a few minor comments

File: .github/workflows/ci.yml (hunk 1 above)
Comment (vittorioromeo) [reply .github/workflows/ci.yml:66]: This one: https://github.com/SFML/SFML/actions/runs/9629381587/job/26558744647#step:17:110

It looks like either complete bogus to me, or something inside the standard library implementation that we cannot really fix. 

Regardless, I cannot reproduce locally with a newer version of MinGW.

File: CMakeLists.txt [hunk 4]
```diff
@@ -204,6 +204,12 @@ if(SFML_ENABLE_SANITIZERS)
     string(APPEND CMAKE_CXX_FLAGS " -fno-omit-frame-pointer -fno-sanitize-recover=all -fsanitize=undefined")
 endif()
 
+option(SFML_ENABLE_STDLIB_ASSERTIONS "Enable standard library assertions" OFF)
+if(SFML_ENABLE_STDLIB_ASSERTIONS)
+    add_compile_definitions(_GLIBCXX_ASSERTIONS=1)                                   # see https://gcc.gnu.org/wiki/LibstdcxxDebugMode
+    add_compile_definitions(_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_EXTENSIVE) # see https://libcxx.llvm.org/Hardening.html
```
Comment (vittorioromeo) [inline CMakeLists.txt:210]: Oh wait @ChrisThrasher, this apparently does not propagate to subdirectories. Should I revert to `add_definitions`?

**Review by ChrisThrasher — APPROVED** [review]

File: CMakeLists.txt (hunk 4 above)
Comment (ChrisThrasher) [reply CMakeLists.txt:210]: Oh really? That's no good. If true then I suppose we have to revert this.

Comment (vittorioromeo) [reply CMakeLists.txt]: Reverted and added a comment.

**Comment by vittorioromeo** [comment]
SFMLBuildMaster: Build this please.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (ChrisThrasher) [inline .github/workflows/ci.yml:66]:
```suggestion
          config: { name: Static with PCH (GCC), flags: -GNinja -DSFML_USE_MESA3D=TRUE -DCMAKE_CXX_COMPILER=g++ -DBUILD_SHARED_LIBS=FALSE -DSFML_ENABLE_PCH=1 -DSFML_ENABLE_STDLIB_ASSERTIONS=FALSE } # disabling stdlib assertions due to false positive
```

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Pull request merged**